src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,414 @@
+
+instance homogeneous_ideal.inhabited : inhabited (homogeneous_ideal 𝒜) :=
+{ default := ⟨⊥, ideal.is_homogeneous.bot _⟩}
+
+instance homogeneous_ideal.has_top :
+  has_top (homogeneous_ideal 𝒜) :=
+⟨⟨⊤, λ _ _ _, by simp only [submodule.mem_top]⟩⟩
+
+@[simp] lemma homogeneous_ideal.eq_top_iff (I : homogeneous_ideal 𝒜) : I = ⊤ ↔ I.1 = ⊤ :=
+subtype.ext_iff
+
+instance homogeneous_ideal.order : partial_order (homogeneous_ideal 𝒜) :=
+partial_order.lift _ subtype.coe_injective
+
+instance homogeneous_ideal.has_mem : has_mem A (homogeneous_ideal 𝒜) :=
+{ mem := λ r I, r ∈ I.1 }
+
+variables {𝒜}
+
+lemma ideal.is_homogeneous.inf {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊓ J).is_homogeneous 𝒜 :=
+λ i r hr, ⟨HI _ hr.1, HJ _ hr.2⟩
+
+lemma ideal.is_homogeneous.Inf {ℐ : set (ideal A)} (h : ∀ I ∈ ℐ, ideal.is_homogeneous 𝒜 I) :
+  (Inf ℐ).is_homogeneous 𝒜 :=
+begin
+  intros i x Hx, simp only [ideal.mem_Inf] at Hx ⊢,
+  intros J HJ,
+  exact h _ HJ _ (Hx HJ),
+end
+
+lemma ideal.is_homogeneous.mul {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I * J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  rw ideal.span_mul_span',
+  refine ⟨s₁ * s₂, congr_arg _ _⟩,
+  exact (set.image_mul (submonoid.subtype _).to_mul_hom).symm,
+end
+
+lemma ideal.is_homogeneous.sup {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊔ J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  refine ⟨s₁ ∪ s₂, _⟩,
+  rw [set.image_union],
+  exact (submodule.span_union _ _).symm,
+end
+
+lemma ideal.is_homogeneous.Sup
+  {ℐ : set (ideal A)} (Hℐ : ∀ (I ∈ ℐ), ideal.is_homogeneous 𝒜 I) :
+  (Sup ℐ).is_homogeneous 𝒜 :=
+begin
+  simp_rw [ideal.is_homogeneous.iff_exists] at Hℐ,
+  set 𝓈 : ℐ → set (homogeneous_submonoid 𝒜) := λ I : ℐ, Exists.some (Hℐ I _) with 𝓈_eq,
+  have h𝓈 : ∀ I : ℐ, I.1 = ideal.span (coe '' 𝓈 I) := λ I : ℐ, Exists.some_spec (Hℐ I _),
+  rw ideal.is_homogeneous.iff_exists,
+  use sUnion (set.range 𝓈),
+  rw [sUnion_range, image_Union, ideal.span, submodule.span_Union],
+  ext r, split,
+  { suffices : Sup ℐ ≤ _, revert r, exact this,
+    rw Sup_le_iff, intros I HI,
+    have ineq1 : I ≤ ideal.span I := ideal.subset_span, refine le_trans ineq1 _,
+    rw ideal.span_le, simp only [coe_subset_coe, ideal.submodule_span_eq], intros x hx,
+    rw submodule.mem_supr, intros J HJ,
+    apply HJ ⟨I, HI⟩, rw ←h𝓈 ⟨I, HI⟩, assumption },
+  { suffices : _ ≤  Sup ℐ, revert r, exact this,
+    rw supr_le_iff, intros I, rw submodule.span_le, intros x hx,
+    simp only [mem_image] at hx, obtain ⟨x', hx1, hx2⟩ := hx,
+    simp only [mem_coe, subtype.val_eq_coe], dsimp only at hx1,
+    apply ideal.mem_Sup_of_mem, use I.2,
+    simp only [subtype.val_eq_coe] at h𝓈 ⊢, rw h𝓈,
+    refine ideal.subset_span _, rw [mem_image], use x', refine ⟨hx1, hx2⟩, },
```

Comment:
Mathlib style is to have one tactic per line when part of a large block like this - can you put a newline between tactics (here and elsewhere)

Reply:
Actually, I golfed away this entire proof using the `choose` tactic.